### PR TITLE
Add OpenTracing for api requests

### PR DESCRIPTION
**Pull request by kschlarman**
**- Summary**
This PR adds OpenTracing for API requests. Currently, the only tracer that is supported is Datadog. The tracer can be enabled and configured using the following env variables, as detailed in the readme:
* `GOTRUE_TRACING_ENABLED`
* `GOTRUE_TRACING_HOST`
* `GOTRUE_TRACING_PORT`
* `GOTRUE_TRACING_TAGS`
* `GOTRUE_SERVICE_NAME`

**- Test plan**
In staging:
* With tracer disabled, verify basic functionality
* WIth tracer enabled, verify request traces show up in datadog w/ appropriate spans based on request.
 

**- Description for the changelog**
Add support for OpenTracing using Datadog tracer

**- A picture of a cute animal (not mandatory but encouraged)**
image

### CODE REVIEW

File: api/tracer.go [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+func newTracingResponseWriter(w http.ResponseWriter) *tracingResponseWriter {
+	return &tracingResponseWriter{w, http.StatusOK}
+}
+
+func (trw *tracingResponseWriter) WriteHeader(code int) {
+	trw.statusCode = code
+	trw.ResponseWriter.WriteHeader(code)
+}
+
+func tracer(next http.Handler) http.Handler {
```
Comment (rybit) [inline api/tracer.go:27]: I think that my only request would be a test of this function

File: api/tracer_test.go [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+	return &testSpan{tags: tt.tags, noopSpan: noopSpan}
+}
+func (tt testTracer) Inject(sm opentracing.SpanContext, format interface{}, carrier interface{}) error {
+	return nil
+}
+func (tt testTracer) Extract(format interface{}, carrier interface{}) (opentracing.SpanContext, error) {
+	return tt.noopTracer.Extract(format, carrier)
+}
+
+type testSpan struct {
```
Comment (kschlarman) [inline api/tracer_test.go:92]: @rybit @mraerino Testing the span tags felt a little cumbersome. Would love suggestions if you see a better way.

File: api/tracer_test.go [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+	assert.Equal(ts.T(), http.StatusNotFound, w.Code)
+	assert.Equal(ts.T(), tags["http.method"], "POST")
+	assert.Equal(ts.T(), tags["http.status_code"], "404")
+	assert.Equal(ts.T(), tags["http.url"], "/something")
+	assert.Equal(ts.T(), tags["resource.name"], "POST /something")
+	assert.NotEmpty(ts.T(), tags["http.request_id"])
+}
+
+type testTracer struct {
+	noopTracer opentracing.NoopTracer
```
Comment (mraerino) [inline api/tracer_test.go:75]: if you make this an embedded struct field you don't have to implement `Inject` and `Extract`

File: api/tracer_test.go [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+}
+func (tt testTracer) Inject(sm opentracing.SpanContext, format interface{}, carrier interface{}) error {
+	return nil
+}
+func (tt testTracer) Extract(format interface{}, carrier interface{}) (opentracing.SpanContext, error) {
+	return tt.noopTracer.Extract(format, carrier)
+}
+
+type testSpan struct {
+	noopSpan opentracing.Span
```
Comment (mraerino) [inline api/tracer_test.go:93]: would also recommend embedding this

File: api/tracer_test.go [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+	ts.API.handler.ServeHTTP(w, req)
+
+	assert.Equal(ts.T(), http.StatusNotFound, w.Code)
+	assert.Equal(ts.T(), spans, []string{"http.handler", "http.handler"})
+}
+
+func (ts *TracerTestSuite) TestTracer_Tags() {
+	tags := map[string]string{}
+
+	tt := &testTracer{tags: &tags, spans: &[]string{}}
```
Comment (mraerino) [inline api/tracer_test.go:59]: this seems a bit overcomplex. why not provide methods on the `testTracer` to get the current list of tags and spans?
that way you also don't have to make them pointers

File: api/tracer_test.go (hunk 2 above)
Comment (rybit) [reply api/tracer_test.go:92]: I think that it is good to try and test the span tags you're using. Datadog is very very picky about what tags are there. If they're not you could break the whole APM setup. 

Also look at https://github.com/netlify/netlify-commons/pull/189/files#diff-04def44a4cd6a9c8549a994b1738c32da5163b39623f56abea1b20ae1ea14428 

Specifically I think you can use `github.com/opentracing/opentracing-go/mocktracer` to cut down on some of the boilerplate

Comment (kschlarman) [reply api/tracer_test.go]: I was looking for something like `mocktracer`, and not sure why I didn't see it 🤦‍♀️ . Thanks for the pointer.

**Review by rybit — APPROVED** [review]

**Pull request merged**